tests/test_core.py

```diff
@@ -220,6 +221,39 @@ def test_add_job_with_none(self):
         with pytest.raises(ValueError):
             self.hass.add_job(None, 'test_arg')
 
+    @patch('asyncio.create_subprocess_exec')
+    def test_restart(self, mock_create):
+        """Check that restart propagates to stop."""
+        process_mock = MagicMock()
+        attrs = {
+            'communicate.return_value': mock_generator(('output', 'error')),
+            'wait.return_value': mock_generator(0)}
+        process_mock.configure_mock(**attrs)
+        mock_create.return_value = mock_generator(process_mock)
+
+        self.hass.start()
+        with patch.object(self.hass, 'async_stop') as mock_stop:
+            self.hass.services.call(ha.DOMAIN, SERVICE_HOMEASSISTANT_RESTART)
+            mock_stop.assert_called_once_with()
+        self.assertEqual(RESTART_EXIT_CODE, self.hass.exit_code)
+
+    @patch('asyncio.create_subprocess_exec')
+    def test_restart_bad_config(self, mock_create):
+        """Check that restart with a bad config doesn't propagate to stop."""
+        process_mock = MagicMock()
+        attrs = {
+            'communicate.return_value':
+                mock_generator((r'\033[hellom'.encode('utf-8'), 'error')),
+            'wait.return_value': mock_generator(1)}
+        process_mock.configure_mock(**attrs)
+        mock_create.return_value = mock_generator(process_mock)
+
+        self.hass.start()
+        with patch.object(self.hass, 'async_stop') as mock_stop:
+            self.hass.services.call(ha.DOMAIN, SERVICE_HOMEASSISTANT_RESTART)
+            mock_stop.assert_not_called()
```

Comment:
Tip for the future: don't use `assert_not_called` because if you would make a typo like calling `mock_stop.assert_not_called_obviously_wrong()` it will pass because it will return a mock object. Better to test explicitly:

```python
assert mock_stop.called
```